File: NQueen/modules/nqueen.py
from random import randint
from os import system

class NQueen:
    def __init__(self, size):
        self.list_queens = list()
        # self.list_queens.append(4)
        # self.list_queens.append(5)
        # self.list_queens.append(6)
        # self.list_queens.append(3)
        # self.list_queens.append(4)
        # self.list_queens.append(5)
        # self.list_queens.append(6)
        # self.list_queens.append(5)
        for i in range(size):
            self.list_queens.append(randint(0, size - 1))

    def get(self, col):
        return self.list_queens[col]

    def put(self, row, col):
        self.list_queens[col] = row

    def pick_up(self, col):
        self.list_queens[col] = -1

    def is_conflict(self, col1, col2):
        if (self.list_queens[col1] == self.list_queens[col2]):
            return True

        if (abs(self.list_queens[col1] - self.list_queens[col2]) == abs(col1 - col2)):
            return True

        return False

    def count_conflicts_at(self, col):
        count = 0
        index = col + 1
        while (index < self.get_size()):
            if (self.is_conflict(col, index)):
                count = count + 1
            index = index + 1
        return count

    def count_conflicts(self):
        count = 0
        col = 0
        while (col < self.get_size()):
            count = count + self.count_conflicts_at(col)
            col = col + 1
        return count

    def get_size(self):
        return len(self.list_queens)

    def print(self, table = True):
        if (table == False):
            print(self.list_queens)
            return

        for i in range(self.get_size()):
            for j in range(self.get_size()):
                if (self.list_queens[j] == i):
                    print('1  ', end='')
                else:
                    print('0  ', end='')
            print()

File: NQueen/src/NQueen.cpp
#include "NQueen.h"

NQueen::NQueen(const int& size)
{
    for (int i = 0; i < size; ++i)
    {
        mQueens.push_back(rand() % size);
    }
}

void NQueen::restart()
{
    for (int i = 0; i < mQueens.size(); ++i)
    {
        mQueens[i] = rand() % mQueens.size();
    }
}
    
void NQueen::setRow(const int& column, const int& row)
{
    mQueens[column] = row;
}

const int& NQueen::getRow(const int& column)
{
    return mQueens[column];
}

bool NQueen::isConflict(const int& firstColumn, const int& secondColumn)
{
    if (mQueens[firstColumn] == mQueens[secondColumn])
    {
        return true;
    }

    if (abs(mQueens[firstColumn] - mQueens[secondColumn]) == abs(firstColumn - secondColumn))
    {
        return true;
    }

    return false;
}

int NQueen::countConflicts()
{
    int count = 0;
    for (int i = 0; i < mQueens.size(); ++i)
    {
        for (int j = i + 1; j < mQueens.size(); ++j)
        {
            if (isConflict(i, j))
            {
                count = count + 1;
            }
        }
    }
    return count;
}

int NQueen::getSize()
{
    return mQueens.size();
}

void NQueen::print()
{
    for (int i = 0; i < mQueens.size(); ++i)
    {
        std::cout << mQueens[i] << ", ";
    }
    
    std::cout << std::endl;
}